### PR TITLE
Dynamic configuration of graph layout forces through blueprints

### LINKED ISSUE #8278: Allow dynamic configuration of forces through blueprints

### PR DESCRIPTION
**Pull request by grtlr**
### Related

* Closes #8278

### What

Title.

### CODE REVIEW

File: crates/store/re_types/definitions/rerun/blueprint/archetypes/force_link.fbs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+namespace rerun.blueprint.archetypes;
+
+/// Aims to achieve a target distance between two nodes that are connected by an edge.
+struct ForceLink (
```
Comment (abey79) [inline crates/store/re_types/definitions/rerun/blueprint/archetypes/force_link.fbs:4]: Shouldn't that be named `ForceEdge` then? We don't use this "link" terminology anywhere else do we?

File: crates/store/re_types/definitions/rerun/blueprint/archetypes/force_collision_radius.fbs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+struct ForceCollisionRadius (
+    "attr.rerun.scope": "blueprint"
+) {
+    /// Whether the force is enabled.
+    enabled: rerun.blueprint.components.Enabled ("attr.rerun.component_optional", nullable, order: 100);
+
+    /// The strength of the force.
+    strength: rerun.blueprint.components.ForceStrength ("attr.rerun.component_optional", nullable, order: 200);
+
+    /// The number of iterations to run the force.
```
Comment (abey79) [inline crates/store/re_types/definitions/rerun/blueprint/archetypes/force_collision_radius.fbs:13]: What does that mean exactly? That the force will be active for the N first iterations of the sim for each time step a collision is "active"? (aka dont hesitate to expend the docstring to help build a better mental model of the force behaviour).

```suggestion
    /// The number of iterations to run the force.
    ///
    /// ...
```

File: crates/store/re_types/definitions/rerun/blueprint/archetypes/force_link.fbs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    "attr.rerun.scope": "blueprint"
+) {
+    /// Whether the force is enabled.
+    enabled: rerun.blueprint.components.Enabled ("attr.rerun.component_optional", nullable, order: 100);
+
+    /// The target distance between two nodes.
+    distance: rerun.blueprint.components.ForceDistance ("attr.rerun.component_optional", nullable, order: 200);
+
+    /// The number of iterations to run the force.
```
Comment (abey79) [inline crates/store/re_types/definitions/rerun/blueprint/archetypes/force_link.fbs:13]: ditto

File: crates/store/re_types/definitions/rerun/blueprint/archetypes/force_many_body.fbs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+/// A force between each pair of nodes that ressembles an electrical charge.
+///
+/// If `ForceStrength` is smaller than 0, it pushes nodes apart, if it is larger than 0 it pulls them together.
+struct ForceManyBody (
+    "attr.rerun.scope": "blueprint"
+) {
+    /// Whether the force is enabled.
+    enabled: rerun.blueprint.components.Enabled ("attr.rerun.component_optional", nullable, order: 100);
+
+    /// The strength of the force.
```
Comment (abey79) [inline crates/store/re_types/definitions/rerun/blueprint/archetypes/force_many_body.fbs:12]:
```suggestion
    /// The strength of the force.
    ///
    /// If `strength` is smaller than 0, it pushes nodes apart, if it is larger than 0 it pulls them together.
```

File: crates/store/re_types/definitions/rerun/blueprint/archetypes/force_position.fbs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+struct ForcePosition (
+    "attr.rerun.scope": "blueprint"
+) {
+    /// Whether the force is enabled.
+    enabled: rerun.blueprint.components.Enabled ("attr.rerun.component_optional", nullable, order: 100);
+
+    /// The strength of the force.
+    strength: rerun.blueprint.components.ForceStrength ("attr.rerun.component_optional", nullable, order: 200);
+
+    /// The position where the nodes should bepulled towards.
```
Comment (abey79) [inline crates/store/re_types/definitions/rerun/blueprint/archetypes/force_position.fbs:13]:
```suggestion
    /// The position where the nodes should be pulled towards.
```

File: crates/store/re_types/src/blueprint/components/enabled_ext.rs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+use re_types_core::datatypes::Bool;
+
+use super::Enabled;
+
+impl Default for Enabled {
+    #[inline]
+    fn default() -> Self {
+        Self(Bool(false))
+    }
+}
```
Comment (abey79) [inline crates/store/re_types/src/blueprint/components/enabled_ext.rs:10]: This is avoidable by adding `Default` to the suitable attribute in the fbs.

File: crates/viewer/re_space_view_graph/src/ui/selection.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
+use re_space_view::{view_property_component_ui, view_property_component_ui_custom};
+use re_types::{
+    blueprint::components::Enabled, Archetype, ArchetypeReflectionMarker, Component as _,
+};
+use re_viewer_context::{ComponentFallbackProvider, SpaceViewId, SpaceViewState, ViewerContext};
+use re_viewport_blueprint::ViewProperty;
+
+pub fn view_property_force_ui<A: Archetype + ArchetypeReflectionMarker>(
```
Comment (abey79) [inline crates/viewer/re_space_view_graph/src/ui/selection.rs:8]: I dont understand this function. What is it doing differently from `view_property_ui`?

**Review by abey79 — APPROVED** [review]
Very nice! And for all its flaws, the default UI isn't that bad already.

Minor quip though: with this PR's web-build, I'm unable to get to a "correct" lattice run:

image

Does it need custom properties to work better? If so that should be added as blueprint in the python version (and would be a bit unfortunate for the rust version).

**Comment by grtlr** [comment]
Thank you for the feedback, I will expand the documentation.

Yeah the web viewer does not resolve the lattice very well. I think once we have a better graph view example (like we talked about earlier), we can demote the lattice to not be part of the manifest anymore.

**Review by Wumpf — APPROVED** [review]
re_component ui, fallback provider & selection ui customization looks good to me!
I wonder if we could do something to make `view_property_force_ui` and similar ones shorter still 🤔 

pity this pr doesn't have a video clip, would surely look great ;)

**Pull request merged**